activerecord/lib/active_record/associations/preloader/through_association.rb

```diff
@@ -18,7 +18,8 @@ def associated_records_by_owner(preloader)
           through_records = owners.map do |owner|
             association = owner.association through_reflection.name
 
-            [owner, Array(association.reader)]
+            center = association.loaded? ? association.target : association.reader
```

Comment:
I think we can extract this logic into a method and use it at the both the places.